### DISCUSSION TITLE
How to document a fixed size array when arguments can be multi typed?

### DISCUSSION THREAD

**Discussion by aaronmallen** [lsegal/yard#1579] [Q&A] [answered]
Based on the information provided [here](https://yardoc.org/types.html) I can specify a fixed sized array with `()` but how do I document the fact that the second argument could be a `String` or a `Symbol`? For example `Array(Integer, <String, Symbol>)` is parsed as "an Array containing (an Integer followed by an Array of (Strings, Symbols))".  What is the correct way to document for this use case?

**Comment by lsegal** [comment, marked as answer]
YARD doesn't really contain a type system, although there are some intuitive conventions used (and parsed by the yard type parser), but these are not enforced anywhere. Because of this, you can ultimately use whatever syntax you prefer, to communicate the type. 

One thing you might consider is splitting this into two types-- specifically:

```ruby
# @param foo [Array(Integer, String), Array(Integer, Symbol)] ...
```

This would be supported by the type parser:

```
an Array containing (an Integer followed by a String);
an Array containing (an Integer followed by a Symbol)
```

You could also use the `|` character commonly used in other type systems to specify this as `Array(Integer, String | Symbol)`, which I think would be relatively understood by consumers, even though the type parser would not technically parse it.

**Reply by aaronmallen** [reply]
Thanks this is helpful. FWIW I'm trying to document a splat that has a mostly definitive shape of exactly N elements in a particular order but some of those arguments could either be `Symbol` or `String`. When I couldn't find an exact match on the yard type parser I just figured I'd ask best practice questions here
